cleverhans/future/tf2/attacks/carlini_wagner_l2.py

```diff
@@ -0,0 +1,210 @@
+    self.MAX_ITERATIONS = max_iterations
+    self.BINARY_SEARCH_STEPS = binary_search_steps
+    self.ABORT_EARLY = abort_early
+    self.CONFIDENCE = confidence
+    self.initial_const = initial_const
+    self.batch_size = batch_size
+    self.data_shape = (batch_size,) + shape
+
+    self.repeat = binary_search_steps >= 10
+
+    self.I_KNOW_WHAT_I_AM_DOING_AND_WANT_TO_OVERRIDE_THE_PRESOFTMAX_CHECK = False
+
+    self.shape = (batch_size, image_size, image_size, num_channels)
+    self.boxmul = (boxmax - boxmin) / 2.
+    self.boxplus = (boxmin + boxmax) / 2.
+
+  def attack(self, imgs, targets):
+    """
+    Perform the L_2 attack on the given images for the given targets.
+
+    If self.targeted is true, then the targets represents the target labels.
+    If self.targeted is false, then targets are the original class labels.
+
+    :param imgs: [batch, width, height, channels]
+    :param labs: e.g. if batch is 1, then labs is like [array([0., 0., 0., 0., 0., 0., 1., 0., 0., 0.])]
+    """
+    r = []
+    print('go up to', len(imgs))
+    for i in range(0, len(imgs), self.batch_size):
+      print('tick', i)
+      r.extend(self.attack_batch(imgs[i:i + self.batch_size], targets))
+    return np.array(r)
+
+  def attack_batch(self, imgs, labs):
+    """
+    Run the attack on a batch of images and labels.
+
+    :param imgs: [batch, width, height, channels]
+    :param labs: e.g. if batch is 1, then labs is like [array([0., 0., 0., 0., 0., 0., 1., 0., 0., 0.])]
+    """
+
+    batch_size = self.batch_size
+
+    def compare(x, y):
+      if not isinstance(x, (float, int, np.int64)):
+        x = x.numpy()
+        x = np.copy(x)
+        if self.TARGETED:
+          x[y] -= self.CONFIDENCE
+        else:
+          x[y] += self.CONFIDENCE
+        x = np.argmax(x)
+      if self.TARGETED:
+        return x == y
+      else:
+        return x != y
+
+    # @tf.function # there are some problems to be fixed, open issues in Tensorflow2.0 beta
+    def train_step(modifier, timg, tlab, const):
+      with tf.GradientTape() as tape:
+        newimg = tf.tanh(modifier + timg) * self.boxmul + self.boxplus
+        output = model.predict(newimg)
+        output = tf.cast(output, dtype=tf.float32)
+        l2dist = tf.reduce_sum(
+            tf.square(newimg - (tf.tanh(timg) * self.boxmul + self.boxplus)), [1, 2, 3])
+        real = tf.math.reduce_sum((tlab) * output, 1)
+        other = tf.math.reduce_max(
+            (1 - tlab) * output - (tlab * 10000), 1)
+        if self.TARGETED:
+          # if targetted, optimize for making the other class most likely
+          loss1 = tf.maximum(0.0, other - real + self.CONFIDENCE)
+        else:
+          # if untargeted, optimize for making this class least likely.
+          loss1 = tf.maximum(0.0, real - other + self.CONFIDENCE)
+
+        loss2 = tf.reduce_sum(l2dist)
+        loss1 = tf.reduce_sum(const * loss1)
+        loss = loss1 + loss2
+      optimizer = optimizers.Adam(self.LEARNING_RATE)
```

Comment:
not entirely familiar with the new OptimizerV2, but I think this line will create a new optimizer at each train step, which would be a bug

Reply:
confirmed with this script:
```
import tensorflow as tf

def train_step(modifier, i):
  x = tf.constant([1., 2, 3], shape=(1,3), dtype=tf.float32)
  c = tf.reduce_sum(10 * x)
  with tf.GradientTape() as tape:
    y = x @ modifier
    loss = tf.norm(c - y)
  optimizer = tf.keras.optimizers.Adam(1)
  grad = tape.gradient(loss, [modifier])
  if i % 5 == 4:
    print(optimizer.iterations)
  optimizer.apply_gradients(zip(grad, [modifier]))
  mmm = optimizer.weights

  return loss, mmm


mod = tf.Variable(tf.ones(shape=(3,1), dtype=tf.float32))
for i in range(15):
  l, m = train_step(mod, i)
```
which returns:
```
<tf.Variable 'iter:0' shape=() dtype=int64, numpy=0>
<tf.Variable 'iter:0' shape=() dtype=int64, numpy=0>
<tf.Variable 'iter:0' shape=() dtype=int64, numpy=0>
```